Workaround broken composed flag for focus events in Firefox 61

web-padawan opened the pull request:
Fixes #275 

I'm using focus monkey-patching to emulate incorrectly working focus, and not setting `composed` flag, so that fallback to the browser default via `composedGetter(this)` could be possible.

When running `npm test` locally, test is failing without the fix and passing with it.

/cc @platosha who is the author of the original focus patch used in `vaadin-grid`
Here is the simplified version of it (without recursive checks) which should be enough.

azakus commented on tests/event-path.html:
```diff
@@ -478,6 +478,43 @@
       el.parentNode.removeChild(el);
     });
 
+    const isFirefox = window.navigator.userAgent.toLowerCase().indexOf('firefox') > -1;
```
Remove this check and do the "broken" focusin and focusout events with a custom event for every browser.

web-padawan replied:
Done. Simulated broken events in test, but note that it now also fails in Chrome without the fix.
So it looks like now I'm testing that patch works, but not the actual Firefox regression. Is that ok to keep it like this?

web-padawan commented on tests/event-path.html:
```diff
@@ -478,6 +478,43 @@
+
+      const spy = sinon.spy();
+      theDiv.addEventListener('focusin', spy);
+      theInput.focus();
```
For some reason, without patching on the prototype I was not able to force shadydom event patching for these mock events. Will continue investigating but would appreciate a help with proper test.

web-padawan commented on src/patch-events.js:
```diff
@@ -137,7 +137,10 @@ let eventMixin = {
+        // TODO(web-padawan): see https://github.com/webcomponents/shadydom/issues/275
+        const brokenComposed = ['focusin', 'focusout'];
+        const nativeComposed = composedGetter(this);
+        this.__composed = brokenComposed.indexOf(this.type) === -1 ? nativeComposed : alwaysComposed[this.type];
```
Done.

web-padawan commented:
Addressed the review feedback. Still don't feel happy with the test, since it fails in all browsers without the fix (ideally, should be only failing in Firefox 61) but it at least confirms the patch.

Feel free to update this PR with the proper test in case if you can find a better solution.

web-padawan commented:
Note: tests are failing on master in Edge 😞 Anything we can do about that?
Will try to debug and find a better solution if possible, but honestly no idea for now.

TimvdLippe commented:
Hm, I think we should revert this for now and get the tests correctly. Could you submit a revert PR @web-padawan ?

The test we write should pass in all browsers, except current Firefox. First we have to verify that. After that, we can make a fix.
